src/sources/futures.rs

```diff
@@ -36,16 +43,69 @@ use crate::{
 /// A future executor as an event source
 #[derive(Debug)]
 pub struct Executor<T> {
-    futures: FuturesUnordered<LocalFutureObj<'static, T>>,
-    new_futures: Channel<LocalFutureObj<'static, T>>,
-    ready_futures: PingSource,
-    waker: Arc<ExecWaker>,
+    /// Shared state between the executor and the scheduler.
+    state: Rc<State<T>>,
+
+    /// Notifies us when the executor is woken up.
+    ping: PingSource,
 }
 
 /// A scheduler to send futures to an executor
 #[derive(Clone, Debug)]
 pub struct Scheduler<T> {
-    sender: Sender<LocalFutureObj<'static, T>>,
+    /// Shared state between the executor and the scheduler.
+    state: Rc<State<T>>,
+}
+
+/// The inner state of the executor.
+#[derive(Debug)]
+struct State<T> {
+    /// The incoming queue of runnables to be executed.
+    incoming: mpsc::Receiver<Runnable<usize>>,
+
+    /// The sender corresponding to `incoming`.
+    sender: Arc<Sender>,
+
+    /// The list of currently active tasks.
+    ///
+    /// This is set to `None` when the executor is destroyed.
+    active: RefCell<Option<Slab<Active<T>>>>,
+}
+
+/// Send a future to an executor.
+///
+/// This needs to be thread-safe, as it is called from a `Waker` that may be on a different thread.
+#[derive(Debug)]
+struct Sender {
+    /// The sender used to send runnables to the executor.
+    sender: Mutex<mpsc::Sender<Runnable<usize>>>,
```

Comment:
Just so that we don't have to wonder about why this mutex is here when we reread this later.

```suggestion
    /// The sender used to send runnables to the executor.
    ///
    /// (`mpsc::Sender` is `!Sync`)
    sender: Mutex<mpsc::Sender<Runnable<usize>>>,
```